pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,208 @@
+					continue
+				}
+
+				secret.Data[kibana.TelemetryFilename] = telemetryBytes
+				if err := r.client.Update(&secret); err != nil {
+					log.Error(err, "failed to update kibana secret")
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (r *Reporter) getResourceUsage() (map[string]interface{}, error) {
+	usage := map[string]interface{}{}
+	for _, f := range []getUsageFn{
+		esUsage,
+		kbUsage,
+		apmUsage,
+		beatUsage,
+		entUsage,
+	} {
+		key, usagePart, err := f(r.client, r.managedNamespaces)
+		if err != nil {
+			return nil, err
+		}
+		usage[key] = usagePart
+	}
+
+	return usage, nil
+}
+
+func esUsage(k8sClient k8s.Client, managedNamespaces []string) (string, interface{}, error) {
+	usage := map[string]int32{resourceCount: 0, podCount: 0}
+
+	var esList esv1.ElasticsearchList
+	for _, ns := range managedNamespaces {
+		if err := k8sClient.List(&esList, client.InNamespace(ns)); err != nil {
+			return "", nil, err
+		}
+
+		for _, es := range esList.Items {
+			usage[resourceCount]++
+			usage[podCount] += es.Status.AvailableNodes
+		}
+	}
+	return "elasticsearches", usage, nil
+}
+
+func kbUsage(k8sClient k8s.Client, managedNamespaces []string) (string, interface{}, error) {
+	usage := map[string]int32{resourceCount: 0, podCount: 0}
+
+	var kbList kbv1.KibanaList
+	for _, ns := range managedNamespaces {
+		if err := k8sClient.List(&kbList, client.InNamespace(ns)); err != nil {
+			return "", nil, err
+		}
+
+		for _, kb := range kbList.Items {
+			usage[resourceCount]++
+			usage[podCount] += kb.Status.AvailableNodes
+		}
+	}
+	return "kibanas", usage, nil
+}
+
+func apmUsage(k8sClient k8s.Client, managedNamespaces []string) (string, interface{}, error) {
+	usage := map[string]int32{resourceCount: 0, podCount: 0}
+
+	var apmList apmv1.ApmServerList
+	for _, ns := range managedNamespaces {
+		if err := k8sClient.List(&apmList, client.InNamespace(ns)); err != nil {
+			return "", nil, err
+		}
+
+		for _, apm := range apmList.Items {
+			usage[resourceCount]++
+			usage[podCount] += apm.Status.AvailableNodes
+		}
+	}
+	return "apms", usage, nil
+}
+
+func beatUsage(k8sClient k8s.Client, managedNamespaces []string) (string, interface{}, error) {
+	typeToName := func(typ string) string { return fmt.Sprintf("%s_count", typ) }
+
+	usage := map[string]int32{resourceCount: 0, podCount: 0}
+	for typ := range beatv1beta1.KnownTypes {
+		usage[typeToName(typ)] = 0
+	}
+
+	var beatList beatv1beta1.BeatList
+	for _, ns := range managedNamespaces {
+		if err := k8sClient.List(&beatList, client.InNamespace(ns)); err != nil {
+			return "", nil, err
+		}
+
+		for _, beat := range beatList.Items {
+			usage[resourceCount]++
+			usage[typeToName(beat.Spec.Type)]++
+			usage[podCount] += beat.Status.AvailableNodes
+		}
+	}
+
+	return "beats", usage, nil
+}
+
+func entUsage(k8sClient k8s.Client, managedNamespaces []string) (string, interface{}, error) {
+	usage := map[string]int32{resourceCount: 0, podCount: 0}
```

Comment:
I was trying to think if we could have a common type for all of these maps so that we can avoid the `interface{}` but could not think of something good rn.

Reply:
I felt this is fairly isolated and allows nesting that might be useful.